### ISSUE TITLE
Consider allowing access::read on function and private storage classes

### ISSUE THREAD

**Issue by jrprice** [gpuweb/gpuweb#1935]
In WGSL today, a `var` declared in the `function` or `private` storage class will always have `read_write` access mode. Since we [limited dynamic indexing on array and matrices](https://github.com/gpuweb/gpuweb/pull/1801), there are now situations where a shader author cannot use `let` to declare these things, and must therefore use a read-write `var`.

We should allow the `access::read` mode on these declarations, so that:
1. We know (more easily) that we can emit `const` (MSL, HLSL) or `NonWritable` (SPIR-V 1.4+)
2. The shader author has a way to prevent themselves from accidentally writing to these variables.

**Comment by jrprice** [comment]
> I don't think you can put `NonWritable` on those variables:

SPIR-V 1.4 expanded `NonWritable` to include objects in the `Private` and `Function` storage classes.

**Comment by alan-baker** [comment]
Seems better to revisit the decision about only static indexes for let declarations.

**Comment by kvark** [comment]
Oh, so this is to support a case like this:
```rust
var<private, read_only> constArray: array<vec4<f32>, 4> = array<vec4<f32>, 4>(
...
);
[[entry(vertex)]]
fn vertex_shader([[builtin(vertex_index)]] index: u32) -> [[builtin(position)]] vec4<f32> {
  return constArray[index];
}
```

It seems to me that this could be an implementation detail to detect that a private isn't modified and emit `NonWritable` on it. We don't have to get this for MVP (but we still could, of course).

**Comment by jrprice** [comment]
Right, that was one of my motivations. I agree that the implementation should be able to infer this, but hints like this can save some analysis.

But I'm probably more interested in the second motivation: protecting shader authors from their own bugs, by allowing them to mark a var as read-only and then producing an error if they try to write to it.

As Alan suggested, if we revisit (and revert) the decision to forbid dynamic indices on non-ref arrays/matrices, then this issue becomes uninteresting since you could always just use a `let` instead. This is definitely my preferred outcome.

**Comment by kvark** [comment]
I don't have any appetite to revisit #1782. It matches our general workflow of providing the intersection of the APIs where polyfills are either numerous or hard. The problem you are trying to solve in this issue is an order of magnitude less important.
